blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,449 @@
+in from Julia, making it a bit awkward to use.   The macro version
+would be much more workable, though it still requires support in the Julia
+parser to recognize expressions like `sqrt.(X)` and `.=`, but the
+decision was made to make fusing the default rather than requiring
+a `@fuse` annotation. In both cases, you still need fast higher-order
+`broadcast`-like functions to support fusion of arbitrary code, as described below.
+
+## Should other languages implement syntactic loop fusion?
+
+Obviously, Julia's approach of syntactic loop fusion relies on the
+fact that, as a young language, we are still relatively free to
+redefine core syntactic elements like `f.(x)` and `x .+ y`.  But
+suppose you were willing to add this or similar syntax to an
+existing language, like Python or Go, or as a DSL on top of those
+languages as suggested above; would you then be able to
+implement the same fusing semantics efficiently?
+
+There is a catch: `2 .* x .+ x .^ 2` is sugar for
+`broadcast(x -> 2*x + x^2, x)` in Julia, but for this to be
+fast we need the [higher-order function](https://en.wikipedia.org/wiki/Higher-order_function)
+`broadcast` to be very fast as well.  First, this
+requires that arbitrary user-defined scalar (non-vectorized!) functions like
+`x -> 2*x + x^2` be compiled to fast code, which isn't true in
+most other dynamically typed high-level languages like Python, but is usually true
+in statically typed languages like Go.   Second, it ideally requires that
+higher-order functions like `broadcast` be able to [inline](https://en.wikipedia.org/wiki/Inline_expansion)
+the function argument `x -> 2*x + x^2`, and this facility is even
+less common.  (It wasn't available in Julia until version 0.5.)
+
+### The importance of inlining
+
+In particular, consider
+a naive implementation of `broadcast` (only for one-argument functions):
+
+```jl
+function naivebroadcast(f, x)
+    y = similar(x)
+    for i in eachindex(x)
+        y[i] = f(x[i])
+    end
+    return y
+end
+```
+
+In Julia, as in other languages, `f` must be some kind of [function
+pointer](https://en.wikipedia.org/wiki/Function_pointer) or [function
+object](https://en.wikipedia.org/wiki/Function_object). Normally, a call
+`f(x[i])` to a function object `f` must figure out where the actual [machine
+code](https://en.wikipedia.org/wiki/Machine_code) for the function is (in Julia,
+this involves dispatching on the type of `x[i]`; in object-oriented languages,
+it might involve dispatching on the type of `f`), push the argument `x[i]`
+etcetera to `f` via a register and/or a [call stack](https://en.wikipedia.org/wiki/Call_stack),
+jump to the machine instructions to execute them, jump back to
+the caller `naivebroadcast`, and extract the return value.
+That is, calling a function argument `f` involves some overhead beyond
+the cost of the computations inside `f`.
+
+If `f(x)` is expensive enough, then the overhead of the function call may be negligible,
```

Comment:
It also prevents SIMD vectorization, which is a big deal for standard arithmetic operations.